src/modalities/dataloader/create_packed_data.py

```diff
@@ -68,54 +85,98 @@ def run(self, dst_path: Path = None):
+            os.environ["TOKENIZERS_PARALLELISM"] = "false"
+            self._launch_parallelized_workers(dst_path)
+        finally:
+            os.unsetenv("TOKENIZERS_PARALLELISM")
+
+        if self._exception_buffer:
+            raise self._exception_buffer[0]
+
+    def _launch_parallelized_workers(self, dst_path: Path):
+        writer = multiprocessing.Process(target=self._writer_thread(dst_path))
+        writer.start()
+        processor_threads = [
+            multiprocessing.Process(target=self._process_thread, args=(i,)) for i in range(self._number_of_processes)
+        ]
+        for p in processor_threads:
+            p.start()
+        for p in processor_threads:
+            p.join()
+        self._stop_processing()
+        writer.join()
+
+    def _stop_processing(self):
+        self._tokens_to_get_written.put(None)
+
+    def _generator_for_tokens_to_get_written(self):
+        while True:
+            self._check_for_parallel_errors()
+            tokens = self._tokens_to_get_written.get()
+            if tokens is None:
+                break
+            yield tokens
+
+    def _check_for_parallel_errors(self):
+        if self._exception_buffer:
+            raise RuntimeError("Exception found in exception buffer. Another thread encountered a problem apparently.")
+
+    def _writer_thread(self, dst_path: Path) -> Callable:
+        def writer():
+            index_list = []
+            with dst_path.open("wb") as f:
+                # allocate first self.header_size_in_bytes bytes for header (encodes length of data section)
+                # not possible to prepend header after determining size of data section
+                f.write((0).to_bytes(self.DATA_SECTION_LENGTH_IN_BYTES, byteorder="big"))
+                f.write(self._token_size_in_bytes.to_bytes(self.TOKEN_SIZE_DESCRIPTOR_LENGTH_IN_BYTES, byteorder="big"))
+                curr_offset = self.HEADER_SIZE_IN_BYTES
+
+                # write data section (tokens)
+                for tokens_as_bytes in tqdm(
+                    self._generator_for_tokens_to_get_written(), desc="Processed Samples", total=len(self._reader)
+                ):
+                    f.write(tokens_as_bytes)
+                    segment_length = len(tokens_as_bytes)
+                    index_list.append((curr_offset, segment_length))
+                    curr_offset += segment_length
+
+                # write index
+                f.write(pickle.dumps(index_list))
+
+            self._update_data_length_in_pre_allocated_header(dst_path, index_list)
+
+        return writer
+
+    def _process_thread(self, process_id: int):
+        self._check_for_parallel_errors()
+        for idx in range(process_id, len(self._reader), self._number_of_processes):
+            line = self._reader[idx]
+            try:
+                self._tokens_to_get_written.put(self._process_line(line))
```

Comment:
You are right. The order is not kept here. Implementationwise and algorithmically, this does not really matter, since the packing works on the "continuous" stream of tokens in the pbin file.
However if a certain order is injected into the original data, this might be an issue and difficult if not even impossible to undo.